ios/amfi/amfi.go

```diff
@@ -0,0 +1,174 @@
+	if err != nil {
+		return fmt.Errorf("EnableDevMode: failed decoding response from service with err: %w", err)
+	}
+
+	plist, err := ios.ParsePlist(responseBytes)
+	if err != nil {
+		return fmt.Errorf("EnableDevMode: failed parsing response plist with err: %w", err)
+	}
+
+	// Check if we have an error returned by the service
+	if _, ok := plist["Error"]; ok {
+		return fmt.Errorf("EnableDevMode: could not enable developer mode through amfi service")
+	}
+
+	if _, ok := plist["success"]; ok {
+		return nil
+	}
+
+	return fmt.Errorf("EnableDevMode: could not enable developer mode through amfi service but no error or success was reported")
+}
+
+// When you enable developer mode and device is rebooted, you get a popup on the device to finish enabling developer mode
+// This function "accepts" that popup
+func (devModeConn *Connection) EnableDevModePostRestart() error {
+	reader := devModeConn.deviceConn.Reader()
+
+	request := map[string]interface{}{"action": 2}
+
+	bytes, err := devModeConn.plistCodec.Encode(request)
+	if err != nil {
+		return fmt.Errorf("EnableDevModePostRestart: failed encoding request to service with err: %w", err)
+	}
+
+	err = devModeConn.deviceConn.Send(bytes)
+	if err != nil {
+		return fmt.Errorf("EnableDevModePostRestart: failed sending request bytes to service with err: %w", err)
+	}
+
+	responseBytes, err := devModeConn.plistCodec.Decode(reader)
+	if err != nil {
+		return fmt.Errorf("EnableDevModePostRestart: failed decoding response from service with err: %w", err)
+	}
+
+	plist, err := ios.ParsePlist(responseBytes)
+	if err != nil {
+		return fmt.Errorf("EnableDevModePostRestart: failed parsing response plist with err: %w", err)
+	}
+
+	if _, ok := plist["success"]; ok {
+		return nil
+	}
+
+	return fmt.Errorf("EnableDevModePostRestart: could not enable developer mode post restart through amfi service")
+}
+
+func EnableDeveloperMode(device ios.DeviceEntry, enablePostRestart bool) error {
+	// Don't try to enable if it already is
+	devModeEnabled, err := imagemounter.IsDevModeEnabled(device)
+	if err != nil {
+		return fmt.Errorf("EnableDeveloperMode: failed checking developer mode status with err: %w", err)
+	}
+
+	if devModeEnabled {
+		log.Info("Developer mode is already enabled on the device")
+		return nil
+	}
+
+	// Perform the first step of developer mode enablement and wait for the device to restart
+	conn, err := New(device)
+	if err != nil {
+		return fmt.Errorf("EnableDeveloperMode: failed connecting to amfi service with err: %w", err)
+	}
+
+	err = conn.EnableDevMode()
+	if err != nil {
+		return fmt.Errorf("EnableDeveloperMode: failed enabling developer mode with err: %w", err)
+	}
+	log.Infof("Successfully enabled developer mode on device `%s`, device will restart", device.Properties.SerialNumber)
+
+	udid := device.Properties.SerialNumber
+	log.Info("Waiting for device to restart after enabling developer mode")
+	ticker := time.NewTicker(10 * time.Second)
```

Comment:
10s per default seems pretty long. And it wouldn't hurt to do this more frequently, the only thing that happens is the `device, err = ios.GetDevice(udid)`, which is pretty cheap.

Reply:
Actually initially I used 5s, the problem is that the device takes some time to reboot after firing the dev mode request and it was sometimes getting the device and proceeding before it was even rebooted 😓 😅